safe_transaction_service/history/models.py

```diff
@@ -499,41 +500,45 @@ def to_erc721_transfer(self):
 
 
 class ERC721TransferManager(TokenTransferManager):
-    # TODO Optimize this
-    def erc721_owned_by(self, address: str) -> List[Tuple[str, int]]:
+    def erc721_owned_by(
+        self, address: ChecksumAddress
+    ) -> List[Tuple[ChecksumAddress, int]]:
         """
         Returns erc721 owned by address, removing the ones sent
 
         :return: List of tuples(token_address: str, token_id: int)
         """
-        # Get all the token history
-        erc721_events = self.to_or_from(address)
-        # Get tokens received and remove tokens transferred
-        tokens_in: Tuple[str, int] = []
-        tokens_out: Tuple[str, int] = []
-        for erc721_event in erc721_events:
-            token_address = erc721_event.address
-            token_id = erc721_event.token_id
-            if token_id is None:
-                logger.error(
-                    "TokenId for ERC721 info token=%s with owner=%s can never be None",
-                    token_address,
-                    address,
-                )
-                continue
-            if erc721_event.to == erc721_event._from:
-                continue  # Nice try ¯\_(ツ)_/¯
-
-            if erc721_event.to == address:
-                list_to_append = tokens_in
-            else:
-                list_to_append = tokens_out
-            list_to_append.append((token_address, token_id))
 
-        for token_out in tokens_out:  # Remove tokens sent from list
-            if token_out in tokens_in:
-                tokens_in.remove(token_out)
-        return tokens_in
+        with connection.cursor() as cursor:
+            hex_address = HexBytes(address)
+            # Queries all the ERC721 IN and all OUT and only returns the ones currently owned
+            cursor.execute(
+                """
+                SELECT Q1.address, Q1.token_id
+                FROM (SELECT address,
+                             token_id,
+                             Count(*) AS count
+                      FROM   history_erc721transfer
+                      WHERE  "to" = %s AND "to" != "_from"
+                      GROUP  BY address,
+                                token_id) Q1
+                     LEFT JOIN (SELECT address,
+                                       token_id,
+                                       Count(*) AS count
+                                FROM   history_erc721transfer
+                                WHERE  "_from" = %s AND "to" != "_from"
+                                GROUP  BY address,
+                                          token_id) Q2
+                            ON Q1.address = Q2.address
+                               AND Q1.token_id = Q2.token_id
+                WHERE Q1.count > COALESCE(Q2.count, 0)
+                """,
+                [hex_address, hex_address],
+            )
+            return [
```

Comment:
What would it happen if there is any change on the database model? Would be detected by unit test to change the query acordingly?

Reply:
Yes

Reply:
Ok, I would like to suggest catch the exception (if model changes) and mange it using orm queries like a second chance. What do you think?

Reply:
I don't agree. If model changes we will notice due to the tests. If we do as you propose and model changes, tests will keep passing and we will not notice that we are using a suboptimal version of the query

Reply:
I'm not very fond of using raw queries unless it's totally needed, but I'm alright if they are well tested

Reply:
Make sense, I hope that unit test get the change before integration test.

Reply:
Of course. You can try as you don't seem convinced 😉